src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,811 @@
+          // Merge condition:
+          // - x = lhs->index % lhs->upper_factor
+          // - s = lhs->scale / rhs->scale
+          // - c = rhs->lower_factor
+          //
+          //    ((x / (c * s)) * s + (x % (c * s)) / c
+          // => ((x / c) / s) * s + ((x / c) % s)
+          // => (x / c)
+          SplitExprNode* merged = rhs.CopyOnWrite();
+          merged->upper_factor = lhs->upper_factor;
+          // reset args[i] to be zero.
+          lhs.CopyOnWrite()->scale = 0;
+          break;
+        }
+      }
+    }
+    // sort by the entry
+    auto fcompare = [](const SplitExpr& lhs, const SplitExpr& rhs) {
+      // order by scale first
+      if (lhs->scale > rhs->scale) return true;
+      if (lhs->scale < rhs->scale) return false;
+      // then order by factor
+      if (lhs->lower_factor > rhs->lower_factor) return true;
+      if (lhs->lower_factor < rhs->lower_factor) return false;
+      // then order by upper factor
+      if (lhs->upper_factor > rhs->upper_factor) return true;
+      if (lhs->upper_factor < rhs->upper_factor) return false;
+      // tie.
+      return false;
+    };
+    std::stable_sort(args.begin(), args.end(), fcompare);
+  }
+  /*!
+   * \brief Return the normal Expr that is equivalent to self.
+   * \return The normal expression.
+   */
+  Expr Normalize() const final {
+    this->SimplifySplitExprs();
+    Type dtype = this->type;
+    // quick path 1.
+    if (this->args.size() == 0) {
+      return make_const(dtype, this->base);
+    }
+    // Positive scales first
+    Expr res = make_const(dtype, 0);
+    for (size_t i = 0; i < this->args.size(); ++i) {
+      if (this->args[i]->scale > 0) {
+        res = res + this->args[i]->Normalize();
+      }
+    }
+    if (this->base > 0) {
+      res = res + make_const(dtype, this->base);
+    }
+    // negative scales follows using sub.
+    for (size_t i = 0; i < this->args.size(); ++i) {
+      if (this->args[i]->scale < 0) {
+        res = res - this->args[i]->NormalizeWithScale(-1);
+      }
+    }
+    if (this->base < 0) {
+      res = res - make_const(dtype, -this->base);
+    }
+    return res;
+  }
+  /*!
+   * \brief Whether self is divisible by scale.
+   * \param scale The scale to be applied.
+   */
+  bool DivisibleBy(int64_t scale) {
+    if (base % scale != 0) return false;
+    for (size_t i = 0; i < this->args.size(); ++i) {
+      if (args[i]->scale % scale != 0) return false;
+    }
+    return true;
+  }
+  /*!
+   * \brief mul scale to self.
+   * \param scale The scale to be applied.
+   */
+  void MulToSelf(int64_t scale) {
+    this->base *= scale;
+    for (size_t i = 0; i < this->args.size(); ++i) {
+      args[i].CopyOnWrite()->scale *= scale;
+    }
+  }
+  /*!
+   * \brief divide by scale.
+   * \param scale The scale to be applied.
+   */
+  void DivideBy(int64_t scale) {
+    this->base /= scale;
+    for (size_t i = 0; i < this->args.size(); ++i) {
+      args[i].CopyOnWrite()->scale /= scale;
```

Comment:
Shouldn't we raise an error if some of the scales are not divisible by the argument?